External dojo deps

maier49 opened the pull request:
**Type:** feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [x] Unit or Functional tests are included in the PR

<!--
Our bots should ensure:

* [ ] All contributors have signed a CLA
* [ ] The PR passes CI testing
* [ ] Code coverage is maintained
* [ ] The PR has been reviewed and approved
-->

**Description:**

Partially addresses #165, by providing a mechanism for integrating Dojo 1 libraries, or any library that can be consumed with the Dojo 1 loader into a Dojo 2 application.

dylans commented on README.md:
```diff
@@ -87,6 +87,37 @@ Ejecting `@dojo/cli-build-webpack` will produce a `config/build-webpack/webpack.
+### Interop with Dojo 1 libraries
+
+Dojo 1 libraries, whether built or not, can be included in a Dojo 2 application by configuring certain options in the project's `.dojorc` file.
+`.dojorc` is a JSON file that contains configuration for Dojo CLI tasks. Configuration for the `dojo build` task can be provided under the
```
Dojo CLI -> Dojo 2 CLI

dylans commented on README.md:
```diff
@@ -87,6 +87,37 @@ Ejecting `@dojo/cli-build-webpack` will produce a `config/build-webpack/webpack.
+`.dojorc` is a JSON file that contains configuration for Dojo CLI tasks. Configuration for the `dojo build` task can be provided under the
+`build-webpack` property. Dojo 1 dependencies can be specified via a property called `externals` within the `build-webpack` config.
+`externals` is an object and the keys are paths, relative to `node_modules`, specifying the directories or files of the Dojo 1 dependencies that should
+be included. Each value can take three possible types:
```
Each value can take three... value isn't specific enough, also would suggest saying accept rather than take

dylans commented on README.md:
```diff
@@ -87,6 +87,37 @@ Ejecting `@dojo/cli-build-webpack` will produce a `config/build-webpack/webpack.
+`build-webpack` property. Dojo 1 dependencies can be specified via a property called `externals` within the `build-webpack` config.
+`externals` is an object and the keys are paths, relative to `node_modules`, specifying the directories or files of the Dojo 1 dependencies that should
+be included. Each value can take three possible types:
+* `true`: Use this if none of the other options apply
```
None of what other options?

dylans commented on README.md:
```diff
@@ -87,6 +87,37 @@ Ejecting `@dojo/cli-build-webpack` will produce a `config/build-webpack/webpack.
+be included. Each value can take three possible types:
+* `true`: Use this if none of the other options apply
+* An array listing the package names that are in the dependency. This list is needed in order to tell the build process to leave the resolution of these
+files up to Dojo. If any module from a package is imported within application code but not specified in this list, it will fail at runtime.
```
files up to Dojo -> resources to Dojo 1

dylans replied:
fail at runtime -> fail to load at runtime ?

dylans commented on README.md:
```diff
@@ -87,6 +87,37 @@ Ejecting `@dojo/cli-build-webpack` will produce a `config/build-webpack/webpack.
+files up to Dojo. If any module from a package is imported within application code but not specified in this list, it will fail at runtime.
+* An object with the following properties:
+ * `packages`: An array of package names, this array serves the same purpose as providing an array as the value.
+ * `hasLoader`: An optional flag to indicate that the specified layer includes a loader and a separate loader is not needed.
```
layer -> Dojo 1.x layer ?

dylans commented on README.md:
```diff
@@ -87,6 +87,37 @@ Ejecting `@dojo/cli-build-webpack` will produce a `config/build-webpack/webpack.
+ * `packages`: An array of package names, this array serves the same purpose as providing an array as the value.
+ * `hasLoader`: An optional flag to indicate that the specified layer includes a loader and a separate loader is not needed.
+ * `main`: Can be provided if the external dependency is a folder. This is a path within the module to the file to be required. The folder or file
+ itself will be required if this is not specified. This is only relevant if `loadImmediately` or `hasLoader` is `true`.
```
is `true` -> is set to `true`

dylans commented on README.md:
```diff
@@ -87,6 +87,37 @@ Ejecting `@dojo/cli-build-webpack` will produce a `config/build-webpack/webpack.
+ * `hasLoader`: An optional flag to indicate that the specified layer includes a loader and a separate loader is not needed.
+ * `main`: Can be provided if the external dependency is a folder. This is a path within the module to the file to be required. The folder or file
+ itself will be required if this is not specified. This is only relevant if `loadImmediately` or `hasLoader` is `true`.
+ * `to`: An alternate location for a dependency to be copied to. The location is relative to the `externals`
```
An alternate location for a dependency to be copied to -> When copying a dependency, this specifies an alternate target location

dylans commented on README.md:
```diff
@@ -87,6 +87,37 @@ Ejecting `@dojo/cli-build-webpack` will produce a `config/build-webpack/webpack.
+
+Types for any dependencies included in `externals` can be installed in `node_modules/@types` just like any other dependency.
+
+If the Dojo 1 library in question is a built layer, and has no external dependencies, then the `externals` config is all that's
```
then the `externals` config is all that's -> then `externals` is the only config

dylans commented on README.md:
```diff
@@ -87,6 +87,37 @@ Ejecting `@dojo/cli-build-webpack` will produce a `config/build-webpack/webpack.
+
+If the Dojo 1 library in question is a built layer, and has no external dependencies, then the `externals` config is all that's
+needed to get a working build. However, in some cases not all dependencies are built into the layer file itself. This is often the case for i18n
+resources, with some plugins, if modules are intentionally excluded from a build, if multiple layers have shared dependencies, etc. In this case some
```
with some -> some

dylans replied:
if modules are intentionally excluded from a build -> intentionally excluded modules within a build

dylans replied:
if multiple layers have shared dependencies -> shared dependencies across layers

dylans commented on README.md:
```diff
@@ -87,6 +87,37 @@ Ejecting `@dojo/cli-build-webpack` will produce a `config/build-webpack/webpack.
+resources, with some plugins, if modules are intentionally excluded from a build, if multiple layers have shared dependencies, etc. In this case some
+additional configuration may be needed. To this end, the `externalConfig` property in the `.dojorc` configuration for `build-webpack` can be specified.
+This property will be passed as is to the loader to configure it. The [`dojoConfig` documentation](https://dojotoolkit.org/documentation/tutorials/1.10/dojo_config/)
+elaborates on the available configuration options.
```
available configuration -> available Dojo 1 configuration

dylans commented on src/plugins/ExternalDojoLoaderPlugin.ts:
```diff
@@ -0,0 +1,158 @@
+	 */
+	main?: string;
+	/**
+	 * Optional flag indicating whether this module includes a loader. If no module contains a loader an un-built Dojo
```
un-built Dojo -> un-built Dojo 1

dylans commented on src/plugins/ExternalDojoLoaderPlugin.ts:
```diff
@@ -0,0 +1,158 @@
+	hasLoader?: boolean;
+	/**
+	 * A List of packages defined by this dependency. This is used to populate the externals config. Any packages
+	 * that are defined by this dependency and imported in the application but not included in this list or will not
```
or will not ... or ?

dylans commented on src/plugins/ExternalDojoLoaderPlugin.ts:
```diff
@@ -0,0 +1,158 @@
+	/**
+	 * A List of packages defined by this dependency. This is used to populate the externals config. Any packages
+	 * that are defined by this dependency and imported in the application but not included in this list or will not
+	 * work. A DojoDependency that is an array of strings is equivalent to only defining this property.
```
this property -> this meaning hasLoader?

maier49 commented:
@dylans I believe I've addressed all of your feedback.

matt-gadd commented:
@maier49 so I have a potentially stupid question, if i built my library on a way older version of dojo 1 (silly I know), and now want to include it I but have no loader in the layer. When we include a loader in the build, it will be from the version of `dojo` in cli-build package.json?

matt-gadd commented:
how would we build on this to support more external library types? could this be more generic in terms of it being an AMD externals configuration? (rather than dojo?). or even better decoupling ourselves completely kind of like intern 4 does in terms of it's loading mechanism and config?

maier49 commented:
@matt-gadd To your question about the loader, if you don't include any layer with a loader, then yes you will get the version of `dojo` defined in the package.json. However there's nothing stopping you from building a separate layer that just includes the loader, or bringing along a specific version of `dojo` yourself and using the loader from that.

agubler commented:
I don't think we should be including Dojo as a dependency of cli-build, rather document that if the external needs a loader that it needs to be included specifically (if it's not built into the external itself).

maier49 commented:
@agubler I've removed Dojo as a dependency

@matt-gadd I've reworked this to allow the registration of multiple loaders and left the loader config up to the developer. An example of providing a loader for Dojo 1 modules can be seen in the readme. Unfortunately the way that externals are loaded is pretty baked in, so this basically involved copying an existing plugin and selectively overriding functionality. See [here](https://github.com/webpack/webpack/issues/2280), and [here](https://github.com/webpack/webpack/issues/4171) for discussion in related issues.

kitsonk commented on src/plugins/ExternalLoaderPlugin.ts:
```diff
@@ -0,0 +1,62 @@
+			} ]),
+			[] as { from: string, to: string }[]
+		)));
+		compiler.plugin(<any> 'this-compilation', (compilation: Compilation) => {
```
We should add this to the [WebPack typings](https://github.com/dojo/cli-build/blob/master/src/webpack.d.ts) that are part of the project so we don't have to cast to any here (or even type the `Compilation`).

kitsonk commented on src/plugins/ExternalLoaderPlugin.ts:
```diff
@@ -0,0 +1,62 @@
+import Compiler = require('webpack/lib/Compiler');
+import Compilation = require('webpack/lib/Compilation');
+import Map from '@dojo/shim/Map';
```
Since our support for cli is Node.js 4+ and 4 had `Map` we shouldn't use shim here.  We should instead do like I have done in the [static build PR](https://github.com/dojo/cli-build/pull/172/files#diff-e5e546dd2eb0351f813d63d1b39dbc48) and add the collections lib to `tsconfig.json`

kitsonk commented on src/plugins/ExternalLoaderUmdTemplatePlugin.ts:
```diff
@@ -0,0 +1,308 @@
+import Map from '@dojo/shim/Map';
```
As above, we shouldn't use Map shim with Node.js 4+

kitsonk commented on src/webpack.config.ts:
```diff
@@ -3,7 +3,9 @@ import NormalModuleReplacementPlugin = require('webpack/lib/NormalModuleReplacem
 import { BuildArgs } from './main';
+import { ExternalDep } from './plugins/ExternalLoaderPlugin';
 import Set from '@dojo/shim/Set';
+import Map from '@dojo/shim/Map';
```
And again...

kitsonk commented on tsconfig.json:
```diff
@@ -9,7 +9,7 @@
 		"removeComments": false,
 		"sourceMap": true,
-		"target": "es6",
+		"target": "es5",
```
Was there a reason why we did this?  Just asking, as there likely is a good one.

maier49 replied:
This was because we are now compiling a file that's run in the app `registerLoader`, and Uglify doesn't like es6.

kitsonk replied:
👍

matt-gadd commented:
Just trying to digest this. Is the reason for the copied plugin to support something other than UMD? Because the issues linked seem to be related to bespoke module formats, which I don't think we should bother supporting?

maier49 commented:
@matt-gadd The UMD plugin that I copied and overwrote is responsible for writing the UMD wrapper itself, and it automatically adds calls to `require` external modules.  That could be problematic in the case that, for example, we need to load a layer file before other external dependencies will be available, and couples us to using either global variables or a global `require` as the mechanism for loading external modules.

kitsonk commented on src/plugins/ExternalLoaderUmdTemplatePlugin.ts:
```diff
@@ -167,7 +153,7 @@ export default class ExternalLoaderUmdTemplatePlugin {
 		const orderedExternalModules = keys.reduce((prev, next) => {
 			return prev.concat(customExternalLoaders[next]);
-		}, [] as any[]);
+		}, [] as Array<Loader | Module>);
```
Small nit, but we tend to use as `(Loader | Module)[]` for Array types.

agubler commented on README.md:
```diff
@@ -87,6 +88,45 @@ Ejecting `@dojo/cli-build-webpack` will produce a `config/build-webpack/webpack.
+ | `from` | `string` | false  | A path relative to `node_modules` that tells the loader where it can find this dependency in order to copy it into the built application |
+ | `to` | `string` | true | A path that replaces `from` as the location to copy this dependency to. Dependencies will be copied to `${externalsOutputPath}/${to}` or `${externalsOutputPath}/${from}` if `to` is not provided. |
+ | `name` | `string` | true | Indicates that this path, and any children of this path, should be loaded via the external loader |
+ | `inject` | `string | string[] | boolean` | true | This property indicates that this dependency defines, or includes, scripts or stylesheets that should be loaded on the page. If `inject` is set to `true`, then the file at the location specified by `to` or `from` will be loaded on the page. If this dependency is a folder, and then `inject` can be set to a string or array of strings to define on ...
```
Think that the pipes in the type column are messing with the formatting.

dylans replied:
"and then `inject`" -> "then `inject`" ?

agubler commented on package.json:
```diff
@@ -53,12 +53,14 @@
     "@dojo/has": "next",
     "@dojo/i18n": "next",
     "@dojo/shim": "next",
+    "auto-require-webpack-plugin": "^1.0.1",
```
let's pin this

agubler commented on src/plugins/CoreLoadPlugin.ts:
```diff
@@ -58,7 +58,7 @@ function isContextual(module: NormalModule, issuers: string[]): boolean {
  */
 function stripPath(basePath: string, path: string): string {
-	let resolved = path.replace(basePath + '/', '').replace(/\..*$/, '');
+	let resolved = (basePath ? path.replace(basePath + '/', '') : path).replace(/\..*$/, '');
```
how come we needed this change?

maier49 replied:
When testing the generated ES6 code, the code coverage generation is actually dropping the default value `''` for basePath in the object destructuring. That was hiding a bug where if `basePath` was `''` we would remove the leading slash from a path. It was hidden because instead we were calling replace with `'undefined/'`. 

I discovered this when I switched to targeting ES5, although it's back to ES6 now so we would only see this if running tests without coverage.

agubler commented on src/plugins/ExternalLoaderPlugin.ts:
```diff
@@ -0,0 +1,101 @@
+	private _pathPrefix: string;
+	private _hash: boolean;
+
+	constructor(options: {
```
could make a nice type for options and clean the formatting a little in the codez

agubler replied:
Could in destructure in the params, like
```ts
constructor({ externals, outputPath: 'externals', pathPrefix, hash }: Options) {
     // stuff
}
```

agubler commented on src/plugins/ExternalLoaderPlugin.ts:
```diff
@@ -0,0 +1,101 @@
+					return assets.concat(inject.map(path => prefixPath(`${base}/${path}`)));
+				}
+
+				const location = (typeof inject === 'string' && `${base}/${inject}`) || to || from;
```
Think this can be cleaner.

agubler commented on src/plugins/ExternalLoaderPlugin.ts:
```diff
@@ -0,0 +1,101 @@
+ */
+export type ExternalDep = string | ExternalDescriptor;
+
+export default class ExternalDojoLoaderPlugin {
```
Can we drop Dojo from the naming?

dylans commented on README.md:
```diff
@@ -87,6 +88,45 @@ Ejecting `@dojo/cli-build-webpack` will produce a `config/build-webpack/webpack.
 
+### Interop with external libraries
+
+External libraries that can not be loaded normally via webpack can be included in a Dojo 2 providing an implementation of `require` or `define`, and some
```
can not -> cannot

dylans replied:
"can be included in a Dojo 2 providing" missing something here

dylans commented on README.md:
```diff
@@ -87,6 +88,45 @@ Ejecting `@dojo/cli-build-webpack` will produce a `config/build-webpack/webpack.
+
+ | Property | Type | optional | Description |
+ | -------- | ---- | -------- | ----------- |
+ | `from` | `string` | false  | A path relative to `node_modules` that tells the loader where it can find this dependency in order to copy it into the built application |
```
"A path relative to `node_modules` that tells the loader where it can find this dependency in order to copy it into the built application" -> "A path relative to `node_modules` specifying the dependency location to copy into the built application" ?

dylans commented on README.md:
```diff
@@ -87,6 +88,45 @@ Ejecting `@dojo/cli-build-webpack` will produce a `config/build-webpack/webpack.
+ | Property | Type | optional | Description |
+ | -------- | ---- | -------- | ----------- |
+ | `from` | `string` | false  | A path relative to `node_modules` that tells the loader where it can find this dependency in order to copy it into the built application |
+ | `to` | `string` | true | A path that replaces `from` as the location to copy this dependency to. Dependencies will be copied to `${externalsOutputPath}/${to}` or `${externalsOutputPath}/${from}` if `to` is not provided. |
```
"Dependencies will be copied to `${externalsOutputPath}/${to}` or `${externalsOutputPath}/${from}` if `to` is not provided." -> "By default, dependencies will be copied to `${externalsOutputPath}/${to}` or `${externalsOutputPath}/${from}` if `to` is not specified."

dylans commented on README.md:
```diff
@@ -87,6 +88,45 @@ Ejecting `@dojo/cli-build-webpack` will produce a `config/build-webpack/webpack.
+ | `name` | `string` | true | Indicates that this path, and any children of this path, should be loaded via the external loader |
+ | `inject` | `string | string[] | boolean` | true | This property indicates that this dependency defines, or includes, scripts or stylesheets that should be loaded on the page. If `inject` is set to `true`, then the file at the location specified by `to` or `from` will be loaded on the page. If this dependency is a folder, and then `inject` can be set to a string or array of strings to define on ...
+
+ As an example the following config will inject `src/legacy/layer.js` into the application page, declare that modules `a`, `b`, and `c` are external
```
config -> configuration

dylans commented on README.md:
```diff
@@ -87,6 +88,45 @@ Ejecting `@dojo/cli-build-webpack` will produce a `config/build-webpack/webpack.
+ | `inject` | `string | string[] | boolean` | true | This property indicates that this dependency defines, or includes, scripts or stylesheets that should be loaded on the page. If `inject` is set to `true`, then the file at the location specified by `to` or `from` will be loaded on the page. If this dependency is a folder, and then `inject` can be set to a string or array of strings to define on ...
+
+ As an example the following config will inject `src/legacy/layer.js` into the application page, declare that modules `a`, `b`, and `c` are external
+ and should be delegated to the external layer, and then copies over the folder `node_modules/legacy-dep`, from which several files are injected. All of these files will be copied into the `externals` folder, which could be overridden by specifying the `outputPath` property in the `externals` config.
```
"copies over the folder" -> "copy the folder"

dylans replied:
config -> configuration

dylans commented on README.md:
````diff
@@ -87,6 +88,45 @@ Ejecting `@dojo/cli-build-webpack` will produce a `config/build-webpack/webpack.
+ }
+```
+
+Types for any dependencies included in `externals` can be installed in `node_modules/@types` just like any other dependency.
````
"just like" -> ", like"

maier49 commented:
@dylans @agubler I've made updates to address your feedback.

dylans commented on README.md:
```diff
@@ -88,6 +89,45 @@ Ejecting `@dojo/cli-build-webpack` will produce a `config/build-webpack/webpack.
+
+ | Property | Type | optional | Description |
+ | -------- | ---- | -------- | ----------- |
+ | `from` | `string` | `false`  | A path relative to `node_modules` specifying the depndency location to copy into the build application. |
```
depndency -> dependency
